File: backend/beta_auth.py
"""Shared beta credentials; application users are not separate accounts."""
import base64
import binascii
import hmac
import os

from starlette.responses import JSONResponse


def install_beta_auth(app):
    @app.middleware('http')
    async def beta_auth(request, call_next):
        mode = os.environ.get('CONSIGN_BETA_MODE')
        on_vercel = os.environ.get('VERCEL') == '1'
        if mode == '0' or (mode != '1' and not on_vercel):
            response = await call_next(request)
            if on_vercel:
                response.headers['Cache-Control'] = 'no-store'
            return response
        password = os.environ.get('CONSIGN_BETA_PASSWORD', '')
        username = os.environ.get('CONSIGN_BETA_USER', 'beta')
        if not password:
            return JSONResponse({'detail': 'Beta credentials are not configured'}, status_code=503)
        try:
            scheme, value = request.headers.get('authorization', '').split(' ', 1)
            supplied = base64.b64decode(value, validate=True)
            valid = scheme.lower() == 'basic' and hmac.compare_digest(supplied, f'{username}:{password}'.encode())
        except (ValueError, binascii.Error):
            valid = False
        if not valid:
            return JSONResponse({'detail': 'Beta login required'}, status_code=401,
                                headers={'WWW-Authenticate': 'Basic realm="ConsignmentSystem Beta", charset="UTF-8"', 'Cache-Control': 'no-store'})
        response = await call_next(request)
        response.headers['Cache-Control'] = 'no-store'
        return response

File: frontend/lib/beta-auth.ts
import { createHash, timingSafeEqual } from 'node:crypto';

export function requireBetaAuth(request: Request): Response | null {
  const mode = process.env.CONSIGN_BETA_MODE;
  if (mode === '0' || (mode !== '1' && process.env.VERCEL !== '1')) return null;
  const password = process.env.CONSIGN_BETA_PASSWORD;
  if (!password) return new Response('Beta credentials are not configured', { status: 503 });
  const match = /^Basic ([A-Za-z0-9+/]+={0,2})$/i.exec(request.headers.get('authorization') || '');
  const digest = (value: Buffer) => createHash('sha256').update(value).digest();
  const expected = Buffer.from(`${process.env.CONSIGN_BETA_USER || 'beta'}:${password}`);
  if (!match || !timingSafeEqual(digest(Buffer.from(match[1], 'base64')), digest(expected))) {
    return new Response('Beta login required', { status: 401, headers: {
      'WWW-Authenticate': 'Basic realm="ConsignmentSystem Beta", charset="UTF-8"', 'Cache-Control': 'no-store',
    } });
  }
  return null;
}
